### PR TITLE
Fix Pinecone cosine relevance score

### LINKED ISSUE #8207: Issue: Pinecone retriever with Cosine Similarity is treated like Cosine Distance
### Issue you'd like to raise.

When using document search from the existing Pinecone index that was created using Cosine **Similarity** in the `ConversationalRetrievalChain`,  the `score_theshold` would eliminate most relevant documents instead of least relevant ones because the _similarity_ metric will be converted to _distance_.

In [_select_relevance_score_fn](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/pinecone.py#L172) it calls the [_cosine_relevance_score_fn](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/base.py#L169) - which converts the similarity returned from Pinecone search to distance.

Then, [filtering the documents](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/base.py#L266) based on the `score_threshold` eliminates the most relevant documents instead of least relevant ones.

### Suggestion:

Pinecone subclass should override the `_cosine_relevance_score_fn` to preserve the similarity, since it is what originally comes back from the Pinecone similarity search.

### PR DESCRIPTION
**Pull request by ruiramos**
Fixes: #8207

Description:
Pinecone returns scores (not distances) with cosine similarity. The values according to the docs are [-1, 1], although I could never reproduce negative values.

This PR ensures that the score returned from Pinecone is preserved, rather than inverted, so the most relevant documents can be filtered (eg when using similarity thresholds)

I'll leave this as a draft PR as I couldn't run the tests (my pinecone account might not be enough - some errors were being thrown around namespaces) so hopefully someone who _can_ will pick this up. 

Maintainers:
@rlancemartin, @eyurtsev

### CODE REVIEW

**Review by baskaryan — COMMENTED** [review]
relevance_scores are meant to be refer to similarity, so a relevance score of 0 means dissimilar and score of 1 means similar. if pinecone returns a distance in [-1, 1] then -1 and 1 of default scoring means dissimilarity. so the default _cosine_releance_score_fn of `1 - distance` is almost right, it just needs to take into account absolute values `1 - abs(distance)`, no?

**Comment by baskaryan** [comment]
> Thanks for getting to this so quickly @baskaryan
> 
> I don't think that's the case, according to their docs: [1. distance metrics section](https://docs.pinecone.io/docs/indexes#distance-metrics), [2. blog post on vector similarity](https://www.pinecone.io/learn/vector-similarity/#Cosine-Similarity). This second link states:
> 
> > The cosine similarity is between -1 and 1, where 1 means an angle of 0 (vectors are as close as they can be), 0 means orthogonal, and -1 means that the vectors are pointing in opposite directions.
> 
> So 1 is a perfect match. This matches my experience of using it in practice (noticeably larger numbers for better matches).

i stand corrected! so it's just the cosine of the angle between the two vectors. what i'm not sure about then is how to think about -1 vs 0 — are they equally distant or is -1 "worse"? in which case, should the relevance score be (cos + 1) / 2?

cc @hinthornw any thoughts here?

**Comment by leo-gan** [comment]
@ruiramos Hi , could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

**Comment by leo-gan** [comment]
@baskaryan Could you, please, review it?

File: libs/langchain/langchain/vectorstores/pinecone.py [hunk 1]
```diff
@@ -250,6 +250,14 @@ def _select_relevance_score_fn(self) -> Callable[[float], float]:
                 "(dot product), or euclidean"
             )
 
+    @staticmethod
+    def _cosine_relevance_score_fn(distance: float) -> float:
+        """Pinecone returns cosine distance *scores* between [-1,1]"""
```
Comment (hinthornw) [inline libs/langchain/langchain/vectorstores/pinecone.py:255]:
```
         """Pinecone returns cosine similarity scores between [-1,1]"""
```

I'd change to 

```
 @staticmethod
 def _cosine_relevance_score_fn(score: float) -> float:
        """Pinecone returns cosine distance *scores* between [-1,1]"""
        return (score + 1)/2
```

**Review by hinthornw — APPROVED** [review]

**Pull request merged**